### PR TITLE
WIP: completing PR 56 (add KML parsing and display functionality)

### PR DESCRIPTION
**Pull request by bgilhome**
This is a WIP following on from PR 56 by @axfelix (now closed). I was tasked with completing the migration/refactor of code from an old module islandora_gmap into islandora_simple_map.

As it stands, this PR adds 3 submodules to add extra functionality to map data/parsing:
- islandora_simple_map_scrape_kml
- islandora_simple_map_scrape_ddi
- islandora_simple_map_scrape_mods

It also contains the following updates to the parent module:
- adds hook_islandora_simple_map_get_geojson() and hook_islandora_simple_map_get_geojson_alter(), for modules to return/alter arrays of geojson data to display on object/collection maps
- adds hook_islandora_simple_map_get_kml() and hook_islandora_simple_map_get_kml_alter(), for modules to return/alter arrays of paths to kml data to display on object/collection maps
- invokes above hooks in islandora_simple_map_islandora_view_object_alter() and islandora_simple_map_collection_map_view()
- adds js to display geojson/kml data in google_maps.js

TBD:
- add js to display geojson/kml data in openstreet_maps.js
- DRY the submodule code

NB Islandora is new to me and I'm sure there may be better/clearer module/hook names, I'm open to suggestions.

One issue that I would appreciate some guidance on relates to the ideal parent module/submodule structure. There is some MODS functionality already present in the parent module, from commits cca4c2d and 4fce5ca. AFAICT, the differences between that and the functionality provided by the islandora_simple_map_scrape_kml submodule are:
- parent module doesn't parse MODS if `islandora_simple_map_use_solr` option is true
- parent module filters MODS results by those containing a configurable `islandora_simple_map_coordinate_delimiter` character, default ";", whereas submodule filters by a regex using a "," delimiter. Should these be aligned?
- parent module uses `hook_islandora_simple_map_get_coordinates()` to add MODS data, whereas submodule uses `hook_islandora_simple_map_get_geojson()` and `hook_islandora_simple_map_get_kml()`
- submodule adds options for cleanup & search fallback (using Google Places search)

Given the above, what is the best approach to adding the extra MODS functionality? 2 possible options are:
1. Submodule MODS functionality is merged into parent module
2. Parent module MODS functionality is migrated into submodule

If (1), should the DDI/KML submodule code also live in the parent, along with config options to enable/disable the functionality?

PS apologies for the author name/email changes in the history from my alternate workstation git configs ('monkii-dev' / 'bryce-gilhome-cp').

### CODE REVIEW

**Comment by whikloj** [comment]
Hey @bgilhome, I'm hoping that @mjordan has more experience with some of this than I. I haven't had a chance to try running this with some test DDI or test KML code so my comments here may suggest paths already attempted, in which case just let me know.

Regarding the existings MODS handling versus the new sub-module.

>   * parent module doesn't parse MODS if islandora_simple_map_use_solr option is true

Yes, as retrieving and parsing MODS can be expensive once you start dealing in collection maps so I think this needs to remain an option.

>   * parent module filters MODS results by those containing a configurable islandora_simple_map_coordinate_delimiter character, default ";", whereas submodule filters by a regex using a "," delimiter. Should these be aligned?

I'll defer to @mjordan on this as it was in-place when I started using the module. I think we could align them, but would need some sort of upgrade that would ensure that the `islandora_simple_map_coordinate_delimiter` character continues to be used for an individual site.

>   * parent module uses hook_islandora_simple_map_get_coordinates() to add MODS data, whereas submodule uses hook_islandora_simple_map_get_geojson() and hook_islandora_simple_map_get_kml()

I am on-board with using GeoJson, KML is more of a Google specific so I see it as a sub-module for those using Google. This raises another interesting option, do we stop with collecting coordinates and instead start migrating to presenting all map features (point, lines, etc) as GeoJson features?

>   * submodule adds options for cleanup & search fallback (using Google Places search)

As this depends on a key for Google, I'd leave this as a configurable sub-module.


> Given the above, what is the best approach to adding the extra MODS functionality? 2 possible options are:
>
> 1. Submodule MODS functionality is merged into parent module
> 1. Parent module MODS functionality is migrated into submodule

The DDI module seems to do 2 things.

1. It grabs specific XPaths from a DDI XML file
1. it generates GeoJson points for these coordinates.

The parent MODS module does the same first part but generates coordinate markers instead of GeoJson Points.

I need to try out Points to see if we could still generate the marker content asynchronously (via AJAX), **but if we can** then we should abstract the parent code to deal with any XML datastream and Xpaths and return the GeoJson points.

Then move that code from the DDI submodule to the main module and make the DDI module simply some configuration around which datastreams to look for and which Xpaths to check.

Or just allow for a pair of values, DSID and array of Xpaths. Then any datastream containing any type of XML could be used.
```
$get_coordinates_from = array(
  'MODS' => array(
      '//mods:mods/mods:subject/mods:coordinates',
  ),
  'DDI' => array(
      '//ddi:some/ddi:element_to/ddi:search',
   ),
);
```

I'll do some cursory code review, but will have to test. Also (as I think you said) the display needs to be completed as only the Google maps stuff is available now.

File: includes/utilities.inc
Comment (whikloj) [inline includes/utilities.inc]: It appears that to display a KML layer on a non-Google map you need to convert it to something else. If that is true then either this should check that you are using Google before it does any work or we need a conversion utilities built in for OSM map providers.

File: islandora_simple_map.module
Comment (whikloj) [inline islandora_simple_map.module]: Not sure why the extra array wrapping is needed here, but if it is it will need to be `array()` as we still support extremely old PHP versions.

File: modules/islandora_simple_map_scrape_ddi/islandora_simple_map_scrape_ddi.module
Comment (whikloj) [inline modules/islandora_simple_map_scrape_ddi/islandora_simple_map_scrape_ddi.module]: 6 is off the end of this URL (I think) so don't you need to validate it exists and is an integer for use above as an index of an array?

File: modules/islandora_simple_map_scrape_ddi/islandora_simple_map_scrape_ddi.module
Comment (whikloj) [inline modules/islandora_simple_map_scrape_ddi/islandora_simple_map_scrape_ddi.module]: These should all start with `ISLANDORA_SIMPLE_MAP_` ie. `ISLANDORA_SIMPLE_MAP_SCRAPE_DDI_KML_XPATHS`. Just to keep naming consistent.

File: modules/islandora_simple_map_scrape_kml/islandora_simple_map_scrape_kml.module
Comment (whikloj) [inline modules/islandora_simple_map_scrape_kml/islandora_simple_map_scrape_kml.module]: Also should start with `ISLANDORA_SIMPLE_MAP_`

File: modules/islandora_simple_map_scrape_mods/includes/admin_form.inc
Comment (whikloj) [inline modules/islandora_simple_map_scrape_mods/includes/admin_form.inc]: This option requires a Google Maps API key, so should have some disabling if we are not using Google maps.

File: modules/islandora_simple_map_scrape_mods/includes/scrape_kml.inc
Comment (whikloj) [inline modules/islandora_simple_map_scrape_mods/includes/scrape_kml.inc]: This seems to be a duplicate of the same file in the DDI module.

File: modules/islandora_simple_map_scrape_mods/islandora_simple_map_scrape_mods.module
Comment (whikloj) [inline modules/islandora_simple_map_scrape_mods/islandora_simple_map_scrape_mods.module]: Here is where you do your semi-colon cleanup.

File: modules/islandora_simple_map_scrape_mods/islandora_simple_map_scrape_mods.module
Comment (whikloj) [inline modules/islandora_simple_map_scrape_mods/islandora_simple_map_scrape_mods.module]: You could probably re-use this function
https://github.com/Islandora-Labs/islandora_simple_map/blob/7.x/includes/utilities.inc#L254-L258

**Review by whikloj — COMMENTED** [review]
Also it would be good to add some tests for the new sub-modules to generate islandora objects and ensure they are parsed/handled correctly to generate the correct output.

**Comment by bgilhome** [comment]
Many thanks for the quick feedback @whikloj, I'll get on to those code cleanup tasks ASAP and wait to hear further from you and @mjordan re: the role of GeoJson within the module/s and what to migrate into the module vs extract into submodule.

**Comment by mjordan** [comment]
@bgilhome Sorry to be absent. I agree with @whikloj that the delimiters used for the same datastream/XPath should be aligned. But, if that inconsistency was in place already, I wouldn't hold up your work by aligning them and adding an update hook. In other words, IMO it's OK to live with the inconsistency for now and someone can tackle it later. A bit of a cop out but I am aware of deadlines to get this work done. Sorry again for taking so long to weigh in.

If I'm completely misunderstanding the question though, let me know.

**Comment by bgilhome** [comment]
Thanks @mjordan and @whikloj for the feedback! I'll get on to this right away and ping this again when ready for review.

**Comment by bgilhome** [comment]
Thanks @whikloj  and @mjordan - perhaps for now then, I'll refactor the MODS submodule functionality (which returns co-ordinate data via GeoJson rather than lat/long pairs) into the parent module by adding an admin option to choose GeoJson or co-ordinates (at least for now, even just for the purpose of testing the ajax functionality as you mentioned @whikloj), and use the existing callbacks/variables for cleanup etc as noted above. If GeoJson later becomes the preferred method, the option can be removed.

The same could be done for the DDI submodule, adding a callback to return lat/long co-ordinates in addition to the GeoJson callback.

The KML functionality could remain in the submodule, as Google Maps dependent, unless/until any OSM plugin is added to handle it (deferred for now).

Does that sound suitable?

**Comment by mjordan** [comment]
@bgilhome it does sound suitable to me, as long as the refactoring doesn't break existing installations when they update the affected modules. In other words, existing configurations should continue to produce existing functionality. Will that be the case?

**Comment by bgilhome** [comment]
@mjordan certainly, all existing functionality will be left as is.

**Comment by mjordan** [comment]
@bgilhome I can give this PR a smoke test when it's ready. If you can briefly outline how I should test, I'd be grateful.

**Comment by bgilhome** [comment]
Hi @mjordan, thanks! I'll write up some instructions with some example data for testing. I'm finishing off moving the DDI functionality to the parent module and then should be ready.

**Comment by mjordan** [comment]
@bgilhome is this ready for testing?

**Comment by bgilhome** [comment]
Hi @mjordan, yep it's ready to test - I've added DDI functionality (and DDI lat/lon xpath selector variables) to the parent module. The KML functionality is all in a submodule.

Existing MODS functionality should behave as before, with an added option to output all scraped data via coordinates (existing behaviour) or GeoJSON. To test, try switching the option, and viewing both object and collection maps (if enabled), using both Google Maps and OpenStreet Maps.

To test the DDI functionality, add a datastream to an object and/or collection, with the ID 'DDI'. Upload an example XML file with lat/lon data (like https://gist.github.com/bgilhome/757b3bf2b0a15aefa8a1214aa129089a), and you should observe the markers added to the map (using either of the 'output' options, and map types, as above).

To test the KML functionality, enable the KML submodule and implement hook_islandora_simple_map_kml_get_kml() in a module, to return an array containing at least one URL to a (publicly-accessible) KML file, e.g. 'https://developers.google.com/maps/documentation/javascript/examples/kml/westcampus.kml' (shows a pin at Google HQ).

I've added a DDI test to the single co-ord unit test, and implementations of the new hooks (hook_islandora_simple_map_get_geojson(), hook_islandora_simple_map_kml_get_kml()) to the hooks tests in the parent module and KML submodule respectively.

Let me know if there are any issues! Thanks.

**Comment by mjordan** [comment]
@bgilhome very good, I'll take a look this evening.

**Comment by mjordan** [comment]
@bgilhome DDI test checks out. :+1: 

Not so much luck with the KML module. I am running on the 56-changes branch and the submodule is enabled. Here's my implementation of the hook in mjtest.module:

```php
/**
 * Implements hook_islandora_simple_map_kml_get_kml().
 */
function mjtest_islandora_simple_map_kml_get_kml(AbstractObject $object) {
  if ($object->id == 'islandora:16') {
    return array(
      'https://developers.google.com/maps/documentation/javascript/examples/kml/westcampus.kml'
    );
  }
}

```

When I view that object, I get the following notice:

> Notice: Undefined offset: 0 in islandora_simple_map_islandora_view_object_alter() (line 185 of /var/www/drupal/sites/all/modules/islandora_simple_map/islandora_simple_map.module).

and the map does not contain a pin:

maps_56_map

I am not using a Google Maps key -  does that matter?

Also, it would be useful to have a brief README for the islandora_simple_map_kml module, explaining how to use it.

**Comment by bgilhome** [comment]
Hi @mjordan, thanks for the testing! I see, there was some code accessing the first element of the coordinates array, however with the new functionality, that array may be empty (if the data contains only GeoJSON or KML). Should be fixed now with an empty() check.

I've added a README to the KML submodule explaining usage and limitations.

Thanks!

**Comment by mjordan** [comment]
PHP Notice is gone, but the map still doesn't show a pin. Is the URL I'm returning correct?

New README looks good, thanks!

**Comment by mjordan** [comment]
Using `dd()`, I can confirm that the value of `$kml` at https://github.com/bgilhome/islandora_simple_map/blob/56-changes/modules/islandora_simple_map_kml/islandora_simple_map_kml.module#L78 has the value

```php
Array
(
    [0] => https://developers.google.com/maps/documentation/javascript/examples/kml/westcampus.kml
)
```

**Comment by bgilhome** [comment]
OK - is it using Google Maps? I'm guessing it probably needs an API key, since it uses another service to parse the KML. I'm looking up the documentation now while it doesn't explicitly say, the example includes an API key. If so I'll add a key check in the KML hook and notes to the README.

**Comment by mjordan** [comment]
Let me try with an API key.

**Comment by bgilhome** [comment]
Thanks! On this note, I've added a todo in the readme re: trying out a KML
parsing library so that local URLs can be used (and without an API key).

**Comment by mjordan** [comment]
Welp, looks like a Google Maps API key is required to get the pin on the KML maps. Here is a map while using a key:

key_yes

But even without a key, I am getting a symbol of some sort on the coordinates' location (Google HQ) and on my own location (Vancouver):

key_no

I didn't notice the no-key behavior earlier, it only seems to happen now that I've used a key (and toggled it off). I could be wrong about that however.

**Comment by mjordan** [comment]
Ha! User error! On the live map without using a Maps API key, those symbols are places that I've saved a photo of. Google's creepy tracking in action.

So, :+1: to all this. Nice work. @whikloj any objections before I merge?

**Comment by whikloj** [comment]
I'd like to test this a bit as we don't use Google Maps and this all seems tied to to that API. Which might mean it will have neither a positive or negative effect. But I will need to confirm.

If you want to merge it, I'll just be careful and do my testing before I pull in any updates.

**Comment by mjordan** [comment]
@whikloj no rush to merge, please test first. I'm reluctant to test on one of our live servers because we're currently doing a bunch of updates and other work.

**Comment by mjordan** [comment]
As per side convo with @whikloj, I'm merging this. Thanks for the work @bgilhome!

**Comment by mjordan** [comment]
Hm, travis is complaining on the PHP 5.3.3 build only about openssl. OK to merge anyway?

File: islandora_simple_map.module
Comment (whikloj) [inline islandora_simple_map.module]: Same re: `[]`

File: islandora_simple_map.module
Comment (whikloj) [inline islandora_simple_map.module]: Same re: `[]`

File: includes/blocks.inc
Comment (whikloj) [inline includes/blocks.inc]: Can't use `[]` arrays as we still 🤷‍♂️ support PHP 5.3.3

File: js/openstreet_maps.js
Comment (whikloj) [inline js/openstreet_maps.js]: Ummm 🙄 should this all be commented out?

File: modules/islandora_simple_map_kml/islandora_simple_map_kml.module
Comment (whikloj) [inline modules/islandora_simple_map_kml/islandora_simple_map_kml.module]: API doc says `hook_islandora_simple_map_kml_get_kml()` returns an array, so this should probably be `return array();`

**Review by whikloj — CHANGES_REQUESTED** [review]
I have left the functional test to @mjordan, this is just things to make PHP 5.3.3 happy and one or two other things I saw

File: islandora_simple_map.module
Comment (bgilhome) [reply islandora_simple_map.module]: Ah yes - sorry, force of habit!

File: js/openstreet_maps.js
Comment (bgilhome) [reply js/openstreet_maps.js]: Yikes, sorry, will uncomment!

**Comment by bgilhome** [comment]
Thanks for picking up those issues @whikloj, I've resolved them and pushed to the branch.

I also added a check in the KML collection function that the Google Maps API option is checked and API key is filled, since without it, no KML layer is added, as you found @mjordan.

**Review by whikloj — APPROVED** [review]

**Comment by whikloj** [comment]
@mjordan ready for you

**Comment by mjordan** [comment]
Done!

**Comment by bgilhome** [comment]
Great news, thanks!

**Pull request merged**
